### PR TITLE
Unity Package: System.Memory.dll version mismatch resolve

### LINKED ISSUE #22251: Error including latest Unity package ("Could not load signature of Google.Protobuf.ByteString:get_Span")
### What version of gRPC and what language are you using?
grpc_unity_package.2.29.0-dev csharp

### What operating system (Linux, Windows,...) and version?
Windows 10

### What runtime / compiler are you using (e.g. python version or version of gcc)
Unity 2019.3.3f1

### What did you do?
I downloaded the last unity package (2.29.0) from https://packages.grpc.io/, extracted it to the Asset/Plugins folder.

### What did you expect to see?
No errors, normal unity editor environment.

### What did you see instead?
Two errors, multiple occurrences of the former, one off the latter:
1) Error: Could not load signature of Google.Protobuf.ByteString:get_Span due to: Could not load file or assembly 'System.Memory, Version=4.0.1.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51' or one of its dependencies. assembly:System.Memory, Version=4.0.1.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51 type:<unknown type> member:(null) signature:<none>

2) Unloading broken assembly Assets/Plugins/Google.Protobuf/lib/net45/Google.Protobuf.dll, this assembly can cause crashes in the runtime

Make sure you include information that can help us debug (full error message, exception listing, stack trace, logs).

See [TROUBLESHOOTING.md](https://github.com/grpc/grpc/blob/master/TROUBLESHOOTING.md) for how to diagnose problems better.

### PR DESCRIPTION
**Pull request by ZCube**
Fix #22251 #23197

System.Memory uses 4.0.1.1 (.net sdk 4.5.3) for grpc and 4.0.1.0 (.net sdk 4.5.2) for protobuf.

Unity3D cannot have two different versions. So change reference from System.Memory 4.0.1.0 to System.Memory 4.0.1.1.

```
@rem System.Memory 4.0.1.0 -> 4.0.1.1 #21908, #22251
@rem WARN: this remove sign from Google.Protobuf.DLL
@rem .netfx 4.6.1 tools tested
"ildasm" /UNICODE /all /text Google.Protobuf.dll /out=Google.Protobuf.il
powershell -Command "(gc Google.Protobuf.il) -replace '.ver 4:0:1:0', '.ver 4:0:1:1' | Out-File -encoding UNICODE Google.Protobuf-mod.il"
@rem need admin permission or Could not create output file, error code=0x80070005
"ilasm.exe" /DLL "%CD%\Google.Protobuf-mod.il" "/output:Google.Protobuf.dll" "/resource=Google.Protobuf.res"
```

@donnadionne

### CODE REVIEW

File: src/csharp/build_unitypackage.bat [hunk 1]
```diff
@@ -72,6 +72,18 @@ copy /Y Grpc.Core\bin\Release\net45\System.Memory.dll unitypackage\unitypackage_
 @rem TODO(jtattermusch): also include XMLdoc
 copy /Y Grpc.HealthCheck\bin\Release\net45\Google.Protobuf.dll unitypackage\unitypackage_skeleton\Plugins\Google.Protobuf\lib\net45\Google.Protobuf.dll || goto :error
 
+@rem System.Memory 4.0.1.0 -> 4.0.1.1 #21908, #22251
```
Comment (jtattermusch) [inline src/csharp/build_unitypackage.bat:75]: Interesting find, but this is a pretty dirty and fragile way of fixing.
Can we use a binding redirect in some way?

Also, we'd only need to ensure that the Google.Protobuf  dll is built with the same SDK as Grpc.Core.dll
(which seems doable). But I'd first like to understand how come Unity cannot handle references to slightly different versions of System.Memory? (this is quite normal on desktop .NET and .NET core, and the Grpc.Core and Google.Protobuf nugets interoperate without a problem).

Comment (ZCube) [reply src/csharp/build_unitypackage.bat:75]: Unity3D has a feature that generates cpp code using a tool called il2cpp. At this time, the conflict occurs for the other two versions of the DLL. So I think Unity3D prevents other versions of the same DLL from being included.

Personally, I would like to modify Protobuf and gRPC to use the same SDK.

gRPC -> 4.5.2 or Protobuf -> 4.5.3

Comment (drew-512) [reply src/csharp/build_unitypackage.bat:75]: ZCube, thanks for your work on driving to get this addressed!

**Comment by jtattermusch** [comment]
Protobuf has been upgraded to the same version of System.Memory as Grpc here: https://github.com/protocolbuffers/protobuf/pull/7940. So once a new version of Google.Protobuf is released, things will work without requiring the workaround from this PR.

Since the workaround from this PR is rather fragile and hacky (as I previously pointed out in the comments), I'm going to close this PR.

**Comment by jtattermusch** [comment]
(also the file being updated by this PR does no longer exist)

**Comment by DanTulovsky** [comment]
Sorry, for the uninitiated, how can I tell when this is fixed? Will a package from https://packages.grpc.io/ fix this at some point?

**Comment by DanTulovsky** [comment]
To answer my question, if you want this to work now, build the Google.Protobuf.dll from source:  https://github.com/protocolbuffers/protobuf/tree/master/csharp

And replace the one in your Unity Assets folder.

**Comment by jtattermusch** [comment]
to clarify my previous statement, the unity package at https://packages.grpc.io/ will have the fix automatically once a new GoogleProtobuf version is released (on nuget.org) and once grpc/grpc codebase bumps its Google.Protobuf dependency.

**Comment by lehmamic** [comment]
I still have this problem, has there already a new version of GoogleProtobuf released? In this case the problem would not have been fixed.

**Pull request closed without merging**